File: 0x10 基本数据结构/0x16 Trie/AcWing143.cpp
#include <iostream>
using namespace std;
const int N = 100006 * 33;
int trie[N][2];

int main() {
  int n;
  cin >> n;
  int ans = 0, t = 1; // t就是tot
  for (int i=1; i<=n; i++) {
    int a;
    scanf("%d", &a);
    // 把每个整数看作长度为31的二进制01串（数值较小时前面补0）
    int p = 1;
    for (int j=31; j>=0; j--) {
      int k = (a >> j) & 1;
      if (!trie[p][k]) trie[p][k] = ++t;
      p = trie[p][k];
    }
    p = 1;
    if (i > 1) {
      int w = 0;
      for (int j=31; j>=0; j--) {
        int k = (a >> j) & 1;
        if (trie[p][k^1]) { // 如果与Ai的当前为相反的字符指针不空
          w = (w << 1) + (k ^ 1);
          p = trie[p][k^1];
        } else { // 反之，空
          w = (w << 1) + k;
          p = trie[p][k];
        }
      }
      ans = max(ans, w ^ a);
    }
  }
  cout << ans << endl;
  return 0;
}

File: 0x10 基本数据结构/0x16 Trie/AcWing143.py
N = 100006*33
trie = [[0]*2 for _ in range(N)]

def main():
    n = int(input())
    ans, t = 0, 1
    aa = [int(x) for x in input().split()]
    for i in range(1, n+1):
        a, p = aa[i-1], 1
        for j in range(31, -1, -1):
            k = (a >> j) & 1
            if not trie[p][k]: t += 1; trie[p][k] = t
            p = trie[p][k]
        p = 1
        if i > 1:
            w = 0
            for j in range(31, -1, -1):
                k = (a >> j) & 1
                if trie[p][k^1]:
                    w = (w << 1) + (k ^ 1)
                    p = trie[p][k^1]
                else:
                    w = (w << 1) + k
                    p = trie[p][k]
            ans = max(ans, w ^ a)
    print(ans)

main()